## Winning Jeopardy

In [1]:
import pandas as pd
import numpy as np
import string

In [2]:
df=pd.read_csv('jeopardy.csv')
df.head()

Show Number    Air Date      Round                         Category  Value  \
0         4680  2004-12-31  Jeopardy!                          HISTORY   $200   
1         4680  2004-12-31  Jeopardy!  ESPN's TOP 10 ALL-TIME ATHLETES   $200   
2         4680  2004-12-31  Jeopardy!      EVERYBODY TALKS ABOUT IT...   $200   
3         4680  2004-12-31  Jeopardy!                 THE COMPANY LINE   $200   
4         4680  2004-12-31  Jeopardy!              EPITAPHS & TRIBUTES   $200   

                                            Question      Answer  
0  For the last 8 years of his life, Galileo was ...  Copernicus  
1  No. 2: 1912 Olympian; football star at Carlisl...  Jim Thorpe  
2  The city of Yuma in this state has a record av...     Arizona  
3  In 1963, live on "The Art Linkletter Show", th...  McDonald's  
4  Signer of the Dec. of Indep., framer of the Co...  John Adams

In [3]:
cols=df.columns
print(cols)

Index(['Show Number', ' Air Date', ' Round', ' Category', ' Value',
       ' Question', ' Answer'],
      dtype='object')


In [4]:
#remove spaces in column names
df.columns=cols.str.replace(' ','')

In [5]:
#function to normalize text(remove punctuation and convert to lowercase)

def NormalizeText(s):
    s=s.lower()
    s=s.translate(str.maketrans('', '', string.punctuation))
    return s
    
clean_question=df['Question'].apply(NormalizeText)
clean_answer=df['Answer'].astype(str).apply(NormalizeText)
df['clean_question']=clean_question
df['clean_answer']=clean_answer

In [6]:
#clean & convert value column

df['Value']=df['Value'].str.replace('$','').str.replace(',','')
df['Value']=pd.to_numeric(df['Value'],errors='coerce')
df['Value'].value_counts()

400.0      42244
800.0      31860
200.0      30455
1000.0     21640
600.0      20377
           ...  
2547.0         1
3599.0         1
1801.0         1
11600.0        1
5001.0         1
Name: Value, Length: 145, dtype: int64

In [7]:
#convert date column to datetime

df['AirDate']=pd.to_datetime(df['AirDate'])
df['AirDate']

0        2004-12-31
1        2004-12-31
2        2004-12-31
3        2004-12-31
4        2004-12-31
            ...    
216925   2006-05-11
216926   2006-05-11
216927   2006-05-11
216928   2006-05-11
216929   2006-05-11
Name: AirDate, Length: 216930, dtype: datetime64[ns]

In order to figure out whether to study past questions, study general knowledge, or not study it all, it would be helpful to figure out two things:

- How often the answer is deducible from the question.
- How often new questions are repeats of older questions

In [8]:
df.head()

ShowNumber    AirDate      Round                         Category  Value  \
0        4680 2004-12-31  Jeopardy!                          HISTORY  200.0   
1        4680 2004-12-31  Jeopardy!  ESPN's TOP 10 ALL-TIME ATHLETES  200.0   
2        4680 2004-12-31  Jeopardy!      EVERYBODY TALKS ABOUT IT...  200.0   
3        4680 2004-12-31  Jeopardy!                 THE COMPANY LINE  200.0   
4        4680 2004-12-31  Jeopardy!              EPITAPHS & TRIBUTES  200.0   

                                            Question      Answer  \
0  For the last 8 years of his life, Galileo was ...  Copernicus   
1  No. 2: 1912 Olympian; football star at Carlisl...  Jim Thorpe   
2  The city of Yuma in this state has a record av...     Arizona   
3  In 1963, live on "The Art Linkletter Show", th...  McDonald's   
4  Signer of the Dec. of Indep., framer of the Co...  John Adams   

                                      clean_question clean_answer  
0  for the last 8 years of his life galileo was u...   copernicus  
1  no 2 1912 olympian football star at carlisle i...   jim thorpe  
2  the city of yuma in this state has a record av...      arizona  
3  in 1963 live on the art linkletter show this c...    mcdonalds  
4  signer of the dec of indep framer of the const...   john adams

In [9]:
def count_matches(row):
    split_answer = row["clean_answer"].split(" ")
    split_question = row["clean_question"].split(" ")
    if "the" in split_answer:
        split_answer.remove("the")
    if len(split_answer) == 0:
        return 0 # so as not to have a division by zero
    match_count = 0
    for item in split_answer:
        if item in split_question:
            match_count += 1
    return match_count / len(split_answer)

df["answer_in_question"] = df.apply(count_matches, axis=1)

In [10]:
df["answer_in_question"].mean()

0.05932504431848426

It appears 5.93% of the times the answer is mentioned in the question.This isn't a huge number, and means that we probably can't just hope that hearing a question will enable us to figure out the answer. We'll probably have to study.

We investigate how often new questions are repeats of older ones.

In [11]:
question_overlap = []
terms_used = set()

df= df.sort_values("AirDate")

for i, row in df.iterrows():
        split_question = row["clean_question"].split(" ")
        split_question = [q for q in split_question if len(q) > 5]
        match_count = 0
        for word in split_question:
            if word in terms_used:
                match_count += 1
        for word in split_question:
            terms_used.add(word)
        if len(split_question) > 0:
            match_count /= len(split_question)
        question_overlap.append(match_count)
df["question_overlap"] = question_overlap

df["question_overlap"].mean()

0.871721036062862

There is about 87% overlap between terms in new questions and terms in old questions. This only looks at a small set of questions, and it doesn't look at phrases, it looks at single terms. This makes it relatively insignificant, but it does mean that it's worth looking more into the recycling of questions

We want to study questions that pertain to high value questions instead of low value questions. This will help earn more money when on Jeopardy.

In [12]:
def determine_value(row):
    value = 0
    if row["Value"] > 800:
        value = 1
    return value

df["high_value"] = df.apply(determine_value, axis=1)

In [13]:
def count_usage(term):
    low_count = 0
    high_count = 0
    for i, row in df.iterrows():
        if term in row["clean_question"].split(" "):
            if row["high_value"] == 1:
                high_count += 1
            else:
                low_count += 1
    return high_count, low_count

In [14]:
from random import choice

terms_used_list = list(terms_used)
comparison_terms = [choice(terms_used_list) for _ in range(10)]

observed_expected = []

for term in comparison_terms:
    observed_expected.append(count_usage(term))

observed_expected

[(0, 1),
 (1, 0),
 (0, 1),
 (1, 0),
 (0, 1),
 (1, 0),
 (0, 1),
 (0, 2),
 (0, 3),
 (0, 1)]

In [15]:
from scipy.stats import chisquare
import numpy as np

high_value_count = df[df["high_value"] == 1].shape[0]
low_value_count = df[df["high_value"] == 0].shape[0]

chi_squared = []
for obs in observed_expected:
    total = sum(obs)
    total_prop = total / df.shape[0]
    high_value_exp = total_prop * high_value_count
    low_value_exp = total_prop * low_value_count
    
    observed = np.array([obs[0], obs[1]])
    expected = np.array([high_value_exp, low_value_exp])
    chi_squared.append(chisquare(observed, expected))

chi_squared

[Power_divergenceResult(statistic=0.3949764642333513, pvalue=0.5296950912486695),
 Power_divergenceResult(statistic=2.5317964247338085, pvalue=0.11157312838169751),
 Power_divergenceResult(statistic=0.3949764642333513, pvalue=0.5296950912486695),
 Power_divergenceResult(statistic=2.5317964247338085, pvalue=0.11157312838169751),
 Power_divergenceResult(statistic=0.3949764642333513, pvalue=0.5296950912486695),
 Power_divergenceResult(statistic=2.5317964247338085, pvalue=0.11157312838169751),
 Power_divergenceResult(statistic=0.3949764642333513, pvalue=0.5296950912486695),
 Power_divergenceResult(statistic=0.7899529284667026, pvalue=0.3741143592744989),
 Power_divergenceResult(statistic=1.184929392700054, pvalue=0.27635474913315955),
 Power_divergenceResult(statistic=0.3949764642333513, pvalue=0.5296950912486695)]

#### Chi-squared results

None of the terms had a significant difference in usage between high value and low value rows. Additionally, the frequencies were all lower than 5, so the chi-squared test isn't as valid. It would be better to run this test with only terms that have higher frequencies

Here are some potential next steps:

Find a better way to eliminate non-informative words than just removing words that are less than 6 characters long. Some ideas:
Manually create a list of words to remove, like the, than, etc.
Find a list of stopwords to remove.
Remove words that occur in more than a certain percentage (like 5%) of questions.
Perform the chi-squared test across more terms to see what terms have larger differences. This is hard to do currently because the code is slow, but here are some ideas:
Use the apply method to make the code that calculates frequencies more efficient.
Only select terms that have high frequencies across the dataset, and ignore the others.
Look more into the Category column and see if any interesting analysis can be done with it. Some ideas:
See which categories appear the most often.
Find the probability of each category appearing in each round.
Use the whole Jeopardy dataset (available here) instead of the subset we used in this mission.
Use phrases instead of single words when seeing if there's overlap between questions. Single words don't capture the whole context of the question well.